### PR TITLE
[Relay] Conv2d grad

### PR DESCRIPTION
**Pull request by vinx13**

### CODE REVIEW

File: src/relay/pass/gradient.cc [hunk 1]
```diff
@@ -348,11 +348,16 @@ struct ReverseAD : ExprMutator {
           args.push_back(ll->Push(VisitExpr(arg)));
         }
         std::vector<Expr> orig_args;
-        for (size_t i = 0; i < args.size(); ++i) {
-          orig_args.push_back(GetValue(op->args[i]->checked_type(), args[i], ll));
+        for (size_t i = 0; i < args.size(); i++) {
```
Comment (MarisaKirisame) [inline src/relay/pass/gradient.cc]: this change make sense, but I had changed gradient. Can you read the change and apply it there? It isnt too hard to understand.

**Comment by MarisaKirisame** [comment]
How is the numerical gradient going?

**Comment by vinx13** [comment]
@MarisaKirisame I disabled numerical tests for now

**Comment by MarisaKirisame** [comment]
dont stress about it. It might be an error with numerical test.

**Comment by SWu** [comment]
I tried playing around with this PR but I'm getting:

```
  File "/usr/tvm/python/tvm/relay/op/_tensor_grad.py", line 257, in conv2d_grad
    data_shape = get_const_tuple(data.checked_type.shape)
  File "/usr/tvm/python/tvm/relay/expr.py", line 47, in checked_type
    raise ValueError("The type checker has not populated"
ValueError: The type checker has not populated the checked_type for this node
```

Is there something I need to call to populate `checked_type`?

**Comment by vinx13** [comment]
@SWu You need to call `InferType` before calling gradient pass

File: src/relay/pass/gradient.cc [hunk 2]
```diff
@@ -271,7 +271,9 @@ Expr LiftTensor(const std::function<Expr(const Expr& t)>& f,
                 LetList* ll) {
   CHECK(IsAtomic(e)) << e;
   if (t.as<TensorTypeNode>()) {
-    return f(e);
+    auto ret = f(e);
+    ret->checked_type_ = t;
+    return ret;
```
Comment (vinx13) [inline src/relay/pass/gradient.cc]: @MarisaKirisame please review this change

**Comment by SWu** [comment]
@vinx13 I am calling InferType pass on the function before gradient pass, but it doesn't seem to make a difference (for what it's worth, I'm using `mode="first_order"` for gradient pass since it seems there are still missing pieces in the relay VM executor for running functions generated by higher order gradient pass). If there's something non-obvious happening here we can discuss elsewhere; I was just curious since it doesn't seem like any other gradients so far require statically knowing the data shapes before calling gradient pass.

**Comment by vinx13** [comment]
@SWu I have updated the gradient pass here, that will copy the checked type before calling the registered gradient function

**Comment by SWu** [comment]
@vinx13 I think I narrowed it down `mode="first_order"` when calling `relay.transform.gradient()`. It seems to work otherwise (using debug executor). I'm not sure if that's a bug in the first_order (as opposed to higher order) gradient transform or somewhere else though.

**Comment by vinx13** [comment]
@MarisaKirisame what do you think is the best way to set the type of `orig` in https://github.com/dmlc/tvm/blob/aee16d87152b4737445e0c8784e9eb7cbc0134c8/src/relay/pass/gradient.cc#L149

**Comment by MarisaKirisame** [comment]
@vinx13 It is indeed the best way. I am pretty ok with it. @jroesch is this ok with you?

**Comment by vinx13** [comment]
@MarisaKirisame The problem is we don't have access to the original call in this function (this is `VisitExpr_(const OpNode*)`

**Comment by MarisaKirisame** [comment]
@vinx13 it is easily solved by requiring the lambda return to take an extra argument. type.
Another thing is the first_order mode shouldnt be used - it support less feature then the higher order mode (even in first order case!) For example, it dont has tuple support as operator call/return/whole function return. I am thinking of deprecating it.

**Comment by junrushao1994** [comment]
If the supporting infra for non first order is not quite ready yet, let's keep first order for now.

**Comment by junrushao1994** [comment]
Also CC @zhiics if you have time

**Comment by zhiics** [comment]
@junrushao1994 Sure. Will do it today.

**Comment by MarisaKirisame** [comment]
@jroesch I want your input with changing the type directly.

**Comment by MarisaKirisame** [comment]
@junrushao1994 @vinx13 can you guys wait a bit before getting this merged? I am a bit hesitant about writing the type directly and want to think about it a bit.

**Comment by junrushao1994** [comment]
@MarisaKirisame not a hurry, we are hesitant as well

**Comment by MarisaKirisame** [comment]
@junrushao1994 I will talk with jared today.

**Comment by MarisaKirisame** [comment]
@junrushao1994 it is ok.

**Comment by zhiics** [comment]
Okay, I think we can now merge. @vinx13 Should we rebase to retrigger the CI since this PR has been here for a while?

**Pull request merged**